rln/src/iden3calc/storage.rs

```diff
@@ -0,0 +1,493 @@
+    let mut ptr = 0usize;
+    w.write_all(WITNESSCALC_GRAPH_MAGIC).unwrap();
+    ptr += WITNESSCALC_GRAPH_MAGIC.len();
+
+    w.write_u64::<LittleEndian>(nodes.len() as u64)?;
+    ptr += 8;
+
+    let metadata = proto::GraphMetadata {
+        witness_signals: witness_signals
+            .iter()
+            .map(|x| *x as u32)
+            .collect::<Vec<u32>>(),
+        inputs: input_signals
+            .iter()
+            .map(|(k, v)| {
+                let sig = proto::SignalDescription {
+                    offset: v.0 as u32,
+                    len: v.1 as u32,
+                };
+                (k.clone(), sig)
+            })
+            .collect(),
+    };
+
+    // capacity of buf should be enough to hold the largest message + 10 bytes
+    // of varint length
+    let mut buf = Vec::with_capacity(metadata.encoded_len() + MAX_VARINT_LENGTH);
+
+    for node in nodes {
+        let node_pb = proto::Node {
+            node: Some(proto::node::Node::from(node)),
+        };
+
+        assert_eq!(buf.len(), 0);
+        node_pb.encode_length_delimited(&mut buf)?;
+        ptr += buf.len();
+
+        w.write_all(&buf)?;
+        buf.clear();
+    }
+
+    metadata.encode_length_delimited(&mut buf)?;
+    w.write_all(&buf)?;
+    buf.clear();
+
+    w.write_u64::<LittleEndian>(ptr as u64)?;
+
+    Ok(())
+}
+
+fn read_message_length<R: Read>(rw: &mut WriteBackReader<R>) -> std::io::Result<usize> {
+    let mut buf = [0u8; MAX_VARINT_LENGTH];
+    let bytes_read = rw.read(&mut buf)?;
+    if bytes_read == 0 {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            "Unexpected EOF",
+        ));
+    }
+
+    let len_delimiter = prost::decode_length_delimiter(buf.as_ref())?;
+
+    let lnln = prost::length_delimiter_len(len_delimiter);
+
+    if lnln < bytes_read {
+        rw.write_all(&buf[lnln..bytes_read])?;
+    }
+
+    Ok(len_delimiter)
+}
+
+fn read_message<R: Read, M: Message + std::default::Default>(
+    rw: &mut WriteBackReader<R>,
+) -> std::io::Result<M> {
+    let ln = read_message_length(rw)?;
+    let mut buf = vec![0u8; ln];
+    let bytes_read = rw.read(&mut buf)?;
+    if bytes_read != ln {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            "Unexpected EOF",
+        ));
+    }
+
+    let msg = prost::Message::decode(&buf[..])?;
+
+    Ok(msg)
+}
+
+pub fn deserialize_witnesscalc_graph(
+    r: impl Read,
+) -> std::io::Result<(Vec<graph::Node>, Vec<usize>, InputSignalsInfo)> {
+    let mut br = WriteBackReader::new(r);
+    let mut magic = [0u8; WITNESSCALC_GRAPH_MAGIC.len()];
+
+    br.read_exact(&mut magic)?;
+
+    if !magic.eq(WITNESSCALC_GRAPH_MAGIC) {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::InvalidData,
+            "Invalid magic",
+        ));
+    }
+
+    let mut nodes = Vec::new();
+    let nodes_num = br.read_u64::<LittleEndian>()?;
+    for _ in 0..nodes_num {
```

Comment:
Probably a one-liner using collect would be more idiomatic

Reply:
Agree.

Reply:
Make sense. 

Btw, do we know the estimate for the number of nodes? I mean that we know the size and it would be nice to allocate memory to the vector at once, instead of copying it every time we push it. 

Like this: 

`let mut nodes = Vec::with_capacity(nodes_num as usize);`

Reply:
> Make sense.
> 
> Btw, do we know the estimate for the number of nodes? I mean that we know the size and it would be nice to allocate memory to the vector at once, instead of copying it every time we push it.
> 
> Like this:
> 
> `let mut nodes = Vec::with_capacity(nodes_num as usize);`

Agree. I will add this.